src/app/GitUI/UserControls/FileStatusSorter.cs

```diff
@@ -0,0 +1,140 @@
+        string b = $"{relativePathB}/";
+        for (int commonEnd = 0; ; ++commonEnd)
+        {
+            if (commonEnd >= a.Length || commonEnd >= b.Length || a[commonEnd] != b[commonEnd])
+            {
+                // Revert possible partial match
+                while (commonEnd > 0 && a[--commonEnd] != '/')
+                {
+                }
+
+                return RelativePath.From(a[..commonEnd]);
+            }
+        }
+    }
+
+    private static TreeNode? GetOrCreateParent(TreeNode previousLeaf, RelativePath currentPath)
+    {
+        RelativePath previousPath = ((FileStatusItem)previousLeaf.Tag).Item.Path;
+        if (previousPath == currentPath && (RelativePath)previousLeaf.Parent.Tag == currentPath)
+        {
+            return previousLeaf.Parent;
+        }
+
+        RelativePath commonPath = GetCommonPath(previousPath, currentPath);
+        if (commonPath.Length == 0)
+        {
+            return null;
+        }
+
+        TreeNode splitCandidate = previousLeaf;
+        RelativePath splitCandidatePath = previousPath;
+        while (splitCandidate.Parent?.Tag is RelativePath path && path.Value.StartsWith(commonPath.Value))
+        {
+            splitCandidate = splitCandidate.Parent;
+            splitCandidatePath = path;
+        }
+
+        if (splitCandidate != previousLeaf && (splitCandidatePath == currentPath || splitCandidatePath == commonPath))
+        {
+            return splitCandidate;
+        }
+
+        return Split(splitCandidate, commonPath);
+    }
+
+    private static TreeNode Split(TreeNode subNode, RelativePath commonPath)
+    {
+        TreeNode parentNode = subNode.Parent ?? throw new ArgumentNullException($"{nameof(subNode)}.{nameof(subNode.Parent)}");
+        int index = parentNode.Nodes.IndexOf(subNode);
+        parentNode.Nodes.Remove(subNode);
+        TreeNode commonFolderNode = new(commonPath.Value) { Tag = commonPath };
+        commonFolderNode.Nodes.Add(subNode);
+        parentNode.Nodes.Insert(index, commonFolderNode);
+        return commonFolderNode;
+    }
+
+    internal static class TestAccessor
+    {
+        public static string GetCommonPath(string a, string b) => FileStatusSorter.GetCommonPath(RelativePath.From(a), RelativePath.From(b)).Value;
+    }
+
+    private sealed class PathFirstComparer : IComparer<GitItemStatus>
+    {
+        public int Compare(GitItemStatus l, GitItemStatus r)
+            => (l, r) switch
+            {
+                (null, null) => 0,
+                (_, null) => -1,
+                (null, _) => 1,
+                (_, _) when l.IsRangeDiff && r.IsRangeDiff => 0,
+                (_, _) when r.IsRangeDiff => -1,
+                (_, _) when l.IsRangeDiff => 1,
+                _ => CompareNonNull(l, r)
+            };
+
+        private static int CompareNonNull(GitItemStatus l, GitItemStatus r)
+        {
+            int pathComparison = (l.Path.Value, r.Path.Value) switch
```

Comment:
The compare previously used the image index to sort, to sort the sub groups.

Reply:
The sorting of groups must have got lost in one of the dozens of rebasing.

Reply:
Better now, but not as I expect it

Order is Both,B,A then by changed,added,removed
I prefer current order changed,added,removed then by both,B,A
That is how I review, can be discussed

changed separates files and submodules now (no specific icon for submodules (my fault), you see both,B,A by the order)
I am divided if the "sub icons for i.e. `+` should be shown at all.
If not then submodules need to be separated, otherwise they could be separated (as they are in the PR)

However, Added/Removed submodule, in added/removed without a specific icon
To see removed only: 2e127404696e215744e1fd64f002eecf8fa15e0b

![image](https://github.com/user-attachments/assets/df897835-fd1c-4e6f-a7b7-a27c407c95b5)

Example where a submodule was removed in one branch, added in another
(and dark icons need to be changed...)
Here there submodules are separated, no top icon

![image](https://github.com/user-attachments/assets/8b7f369c-8c9a-47f8-978e-c5a3a9227258)